core/src/instance.rs

```diff
@@ -235,56 +269,17 @@ impl Instance {
 /// # Panics
 ///
 /// Panics if the channels passed are disconnected.
-pub fn dispatch_action_and_wait(
-    action_channel: &SyncSender<ActionWrapper>,
-    observer_channel: &SyncSender<Observer>,
-    action_wrapper: ActionWrapper,
-) {
-    // Create blocking channel
-    let (sender, receiver) = sync_channel::<()>(1);
-
-    // Create blocking observer
-    let observer_action_wrapper = action_wrapper.clone();
-    let closure = move |state: &State| {
-        if state.history.contains(&observer_action_wrapper) {
-            sender
-                .send(())
-                // the channel stays connected until the first message has been sent
-                // if this fails that means that it was called after having returned done=true
-                .expect("observer called after done");
-            true
+pub fn dispatch_action_and_wait(context: Arc<Context>, action_wrapper: ActionWrapper) {
+    let observer_rx = context.create_observer();
+    dispatch_action(context.action_channel(), action_wrapper.clone());
+
+    loop {
+        if context.state().unwrap().history.contains(&action_wrapper) {
+            return;
         } else {
-            false
+            observer_rx.recv().expect("Local channel must work");
```

Comment:
Could abstract these loops out with something like:

```rust
fn block_with<F>(rx: Receiver<()>, f: F)
where F: FnMut() -> bool {
    loop {
        if f() {
            return
        } else {
            rx.recv().unwrap();
        }
    }
}
```

Reply:
Well, some of those need to match `Some(result)` and some don't. I would actually like to get rid of most of those loops in favor of futures that then use the loop in `Context::block_on`, but as a follow up.